File: rkumquat/src/scanner.rs
use std::any::TypeId;

use crate::{error::error, token::Token, token_type::TokenType};

pub struct Scanner {
    source: String,
    tokens: Vec<Token>,
    start: usize,
    current: usize,
    line: usize,
}

impl Scanner {
    pub fn new(source: String) -> Self {
        Self {
            source,
            tokens: Vec::<Token>::new(),
            start: 0,
            current: 0,
            line: 1,
        }
    }

    pub fn scan_tokens(&mut self) -> &Vec<Token> {
        while !self.is_at_end() {
            self.start = self.current;
            self.scan_token();
        }
        let _ = &self.tokens.push(Token::new(TokenType::Eof, "".to_owned(), None, self.line));
        &self.tokens
    }

    fn scan_token(&mut self) {
        let c = self.advance();
        match c {
            '(' => self.add_token_helper(TokenType::LeftParen),
            ')' => self.add_token_helper(TokenType::RightParen),
            '{' => self.add_token_helper(TokenType::LeftBrace),
            '}' => self.add_token_helper(TokenType::RightBrace),
            ',' => self.add_token_helper(TokenType::Comma),
            '.' => self.add_token_helper(TokenType::Dot),
            '-' => self.add_token_helper(TokenType::Minus),
            '+' => self.add_token_helper(TokenType::Plus),
            ';' => self.add_token_helper(TokenType::Semicolon),
            '*' => self.add_token_helper(TokenType::Star),
            _ => error(self.line, "Unexpected character".to_owned()),
        }
    }

    fn advance(&mut self) -> char {
        self.current += 1;
        self.source.chars().nth(self.current).unwrap()
    }

    fn add_token_helper(&mut self, token_type: TokenType) {
        self.add_token(token_type, None);
    }

    fn add_token(&mut self, token_type: TokenType, literal: Option<TypeId>) {
        let text = &self.source[self.start..self.current];
        self.tokens.push(Token::new(token_type, text.to_owned(), literal, self.line));
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }
}

File: ZKumquat/src/scanner.zig
const TokenType = enum {
    // Single character tokens
    leftParen, rightParen,
    leftBrace, rightBrace,
    comma, dot, minus, plus,
    colon, semicolon, slash, star,

    // One or two character tokens
    bang, bang_equal,
    equal, equal_equal,
    greater, greater_equal,
    less, less_equal,

    // Literals
    identifier, string, number,

    // Keywords
    t_and, t_or, object, interface,
    t_if, t_else, fun, t_for, t_true,
    t_false, nil, print, t_return, super,
    this, t_const, t_while,

    t_error, eof
};

const Token = struct {
    token_type: TokenType,
    start: *u8,
    length: isize,
    line: isize,
};

const Scanner = struct {
    start: *u8,
    current: *u8,
    line: isize,
};

const scanner = Scanner;

fn initScanner(source: []const u8) !void {
    scanner = Scanner {
        .start = source[0],
        .current = source[0],
        .line = 1,
    };
}

fn isAlpha(c: u8) bool {
    return c >= 'a' and c <= 'z'
        or c >= 'A' and c <= 'Z'
        or c == '_';
}

fn isDigit(c: u8) bool {
    return c >= '0' and c <= '9';
}

fn isAtEnd() bool {
    return *scanner.current == '0';
}

fn advance() u8 {
    scanner.current += 1;
    return scanner.current[-1];
}

fn peek() u8 {
    return *scanner.current;
}

fn peekNext() u8 {
    if (isAtEnd()) return '0';
    return scanner.current[1];
}

fn match(expected: u8) bool {
    if (isAtEnd()) return false;
    if (*scanner.current != expected) return false;
    scanner.current += 1;
    return true;
}

//fn makeToken(token_type: TokenType) Token {

//}